modules/javafx.graphics/src/main/java/com/sun/javafx/css/media/MediaQuerySerializer.java

```diff
@@ -0,0 +1,117 @@
+/*
+ * Copyright (c) 2025, Oracle and/or its affiliates. All rights reserved.
+ * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
+ *
+ * This code is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License version 2 only, as
+ * published by the Free Software Foundation.  Oracle designates this
+ * particular file as subject to the "Classpath" exception as provided
+ * by Oracle in the LICENSE file that accompanied this code.
+ *
+ * This code is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+ * version 2 for more details (a copy is included in the LICENSE file that
+ * accompanied this code).
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+
+package com.sun.javafx.css.media;
+
+import com.sun.javafx.css.media.expression.ConjunctionExpression;
+import com.sun.javafx.css.media.expression.ConstantExpression;
+import com.sun.javafx.css.media.expression.FunctionExpression;
+import com.sun.javafx.css.media.expression.NegationExpression;
+import com.sun.javafx.css.media.expression.DisjunctionExpression;
+import javafx.css.StyleConverter;
+import java.io.DataInputStream;
+import java.io.DataOutputStream;
+import java.io.IOException;
+
+/**
+ * Serializes and deserializes a {@link MediaQuery} expression into and from a binary representation.
+ */
+public final class MediaQuerySerializer {
+
+    private MediaQuerySerializer() {}
+
+    /**
+     * The order of {@code QueryType} enum values is important!
+     * It is part of the binary serialization protocol for media query expressions.
+     */
+    private enum QueryType {
+        CONSTANT,
+        FUNCTION,
+        CONJUNCTION,
+        DISJUNCTION,
+        NEGATION;
+
+        static QueryType of(MediaQuery expression) {
+            return switch (expression) {
+                case ConstantExpression _ -> CONSTANT;
+                case FunctionExpression<?> _ -> FUNCTION;
+                case ConjunctionExpression _ -> CONJUNCTION;
+                case DisjunctionExpression _ -> DISJUNCTION;
+                case NegationExpression _ -> NEGATION;
+            };
+        }
+
+        static final QueryType[] VALUES = values();
+    }
+
+    public static void writeBinary(MediaQuery mediaQuery,
+                                   DataOutputStream os,
+                                   StyleConverter.StringStore stringStore) throws IOException {
+        os.writeByte(QueryType.of(mediaQuery).ordinal());
```

Comment:
potential problem:
someone may insert/reorder QueryType enums and break backward compatibility.
at a minimum, a warning comment should be added to QueryType, or perhaps some more reliable code in read/writeBinary?

Reply:
or perhaps use hard-coded values instead 'c', 'f', 'C', 'd', 'n' ?

Reply:
I've added an explicit `serializedId` value.